### PR TITLE
Fix memory widget not stopping thread on keyboard interrupt

### LINKED ISSUE #31351: Fix MemoryWidget not stopping thread on keyboardinterrupt
When you use Ctrl+C for example, to create an KeyboardInterrupt exception on python, it seems to largely be ignored by the Memory Widget's thread for updating, and just continues to create new threads and ignore the keyboard interrupt.

This should be fixed as it breaks the expected behavior for programs on UNIX.

I think it's to do with either 1. The QTimer thread not adequately being stopped when a keyboard interrupt is passed, which leaves it to start ANOTHER QTimer thread that ignores it, until you manage to push an interrupt through when it's in the middle of the python code to update the Widget, or 2. The QTimer is not being adequately told to stop on Workbench exit.

I recommend you review how we handle this on Workbench for Project Recovery as that background thread seems to be tidied up nicely.

### PR DESCRIPTION
**Pull request by Pasarus**
**Description of work.**
Fix memory widget not stopping thread on keyboard interrupt

**To test:**
Test memory widget:
- Run this script 
```
from mantid.simpleapi import *

for ii in range(0, 10000):
    CreateSampleWorkspace(OutputWorkspace=f"ws{ii}")
```
- You should notice that the memory widget's usage is increasing roughly every 2 seconds.

Test this change:
- Close workbench (The process should end quickly and not hang around)
- Open it again
- Perform a KeyboardInturupt on the python script using Ctrl+C or equivalent in a terminal running workbench
- It should not error inside of the memory widget's presenter (check the traceback in the log)
- Try a few more times and just ensure that the memory widget isn't the area it fails inside of.

Fixes #31351  

*This does not require release notes* because **Added during this feature cycle, fixed a bug**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by StephenSmith25 — DISMISSED** [review]
Looks good

File: qt/applications/workbench/workbench/app/mainwindow.py [hunk 1]
```diff
@@ -576,11 +577,18 @@ def closeEvent(self, event):
...
-
-            self.interface_manager.closeHelpWindow()
+            if self.project_recovery is not None:
+                self.project_recovery.stop_recovery_thread()
+                self.project_recovery.closing_workbench = True
+                self.project_recovery.remove_current_pid_folder()
+
+            # Cancel memory widget thread
+            if self.memorywidget is not None:
+                self.memorywidget.presenter.closing_workbench = True
```
Comment (martyngigg) [inline qt/applications/workbench/workbench/app/mainwindow.py:587]: Can we avoid explicitly mentioning workbench? The widget lives in `mantidqt` and should not assume it's part of workbench.
Can the `cancel_memory_update` method not just some internal flag to mark it's closing rather than the outside needing to remember to do that?

**Review by martyngigg — CHANGES_REQUESTED** [review]

File: qt/applications/workbench/workbench/app/mainwindow.py (hunk 1 above)
Comment (Pasarus) [reply qt/applications/workbench/workbench/app/mainwindow.py:587]: I have made it more generic and added it to `cancel_memory_update`.

File: qt/python/mantidqt/widgets/memorywidget/memorypresenter.py [hunk 2]
```diff
@@ -54,5 +69,14 @@ def update_memory_usage(self):
...
-        mem_used_percent, mem_used, mem_avail = get_memory_info()
-        self.view.set_value(mem_used_percent, mem_used, mem_avail)
+        if not self.updating_cancelled:
+            self.thread_on = False
+            mem_used_percent, mem_used, mem_avail = get_memory_info()
+            self.view.set_value(mem_used_percent, mem_used, mem_avail)
+            self._spin_off_another_time_thread()
+
+    def cancel_memory_update(self):
+        self.updating_cancelled = True
```
Comment (martyngigg) [inline qt/python/mantidqt/widgets/memorywidget/memorypresenter.py:79]: Can we have some method documentation, please?

File: qt/python/mantidqt/widgets/memorywidget/memorypresenter.py [hunk 3]
```diff
@@ -25,18 +25,33 @@ class MemoryPresenter(object):
...
         self.view = view
-        self.timer = QTimer()
+        self.timer = Timer(TIME_INTERVAL_MEMORY_USAGE_UPDATE, self.update_memory_usage)
+        self.thread_on = False
+        self.updating_cancelled = False
+        self.application_closing = False
         self.update_memory_usage()
         self.set_bar_color_at_start()
-        self.update_at_regular_intervals()
+        self.start_memory_widget_update_thread()
```
Comment (martyngigg) [inline qt/python/mantidqt/widgets/memorywidget/memorypresenter.py:34]: As we are inside the presenter of the widget perhaps we shorten this to `self.start_update_thread`?

File: qt/python/mantidqt/widgets/memorywidget/memorypresenter.py [hunk 4]
```diff
@@ -25,18 +25,33 @@ class MemoryPresenter(object):
...
-        self.timer.start(TIME_INTERVAL_MEMORY_USAGE_UPDATE)
+        if not self.thread_on:
+            self.timer.start()
+
+    def _spin_off_another_time_thread(self):
+        """
+        Spins off another timer thread, by creating a new Timer thread object and starting it
+        """
+        if not self.thread_on and not self.application_closing:
+            self.timer = Timer(TIME_INTERVAL_MEMORY_USAGE_UPDATE, self.update_memory_usage)
```
Comment (martyngigg) [inline qt/python/mantidqt/widgets/memorywidget/memorypresenter.py:52]: `Timer` is a subclass of Thread so this means that we start a new thread every ~2s? That's quite expensive. Can be do something bit like the interval code [here](https://stackoverflow.com/q/5179467) so we only need to start 1 thread and this thread simply calls the function repeatedly?

Comment (Pasarus) [reply qt/python/mantidqt/widgets/memorywidget/memorypresenter.py]: I have implemented based on that stackoverflow discussion you linked, it "should" be clean and work in other scenarios so I stuck it into mantidqt.utils

File: qt/python/mantidqt/utils/asynchronous.py [hunk 5]
```diff
@@ -215,3 +215,26 @@ def __str__(self):
...
+        def wrap(*args, **kwargs):
+            stop = threading.Event()
+
+            def inner_wrap():
+                function(*args, **kwargs)
+                while not stop.isSet():
+                    stop.wait(interval)
+                    function(*args, **kwargs)
+            t = threading.Timer(0, inner_wrap)
+            # t.daemon = True
```
Comment (martyngigg) [inline qt/python/mantidqt/utils/asynchronous.py:236]: May I ask why this is commented out?

File: qt/python/mantidqt/widgets/memorywidget/memorypresenter.py [hunk 6]
```diff
@@ -7,36 +7,29 @@
...
-    This happens at the beginning as well as every
-    TIME_INTERVAL_MEMORY_USAGE_UPDATE (ms) using QTimer
-    Besides, sets the style of the memory(progress) bar
-    at the beginning
+    Gets system memory usage information and passes it to the memory view this happens at the beginning as well as
+    every TIME_INTERVAL_MEMORY_USAGE_UPDATE (s) using threading.Timer; Also, sets the style of the memory(progress) bar
+    on construction.
     """
     def __init__(self, view):
+        import pydevd_pycharm
```
Comment (martyngigg) [inline qt/python/mantidqt/widgets/memorywidget/memorypresenter.py:25]: Some debugging statements left I think :)

File: qt/python/mantidqt/utils/asynchronous.py (hunk 5 above)
Comment (Pasarus) [reply qt/python/mantidqt/utils/asynchronous.py]: Good catch!

File: qt/python/mantidqt/widgets/memorywidget/memorypresenter.py (hunk 6 above)
Comment (Pasarus) [reply qt/python/mantidqt/widgets/memorywidget/memorypresenter.py]: Ack I un-stashed the changes as I was comparing the previous implementation and this one, and just committed. Thanks for catching that!

File: qt/applications/workbench/workbench/app/mainwindow.py [hunk 7]
```diff
@@ -546,21 +546,22 @@ def arrange_layout(self, layout):
...
+            if not self.project.saved:
+                # Offer save
+                if self.project.offer_save(self):
+                    # Cancel has been clicked
+                    event.ignore()
+                    return
+
         # Close editors
-        if self.editor.app_closing():
+        if self.editor is None or self.editor.app_closing():
```
Comment (martyngigg) [inline qt/applications/workbench/workbench/app/mainwindow.py:564]: Could I ask why this extra check for the editor being None is required?

Comment (Pasarus) [reply qt/applications/workbench/workbench/app/mainwindow.py:564]: In test_mainwindow, some tests it is possible for it to be None, previously theses tests didn't clean up properly as they didn't call close() on tearDown.

File: qt/python/mantidqt/widgets/memorywidget/test/test_memorypresenter.py [hunk 8]
```diff
@@ -42,8 +48,10 @@ def test_presenter(self):
         self.assertEqual(self.presenter.view.set_value.call_count, 1)
         self.assertEqual(self.presenter.view.set_bar_color.call_count, 1)
 
-        self.presenter.update_memory_usage()
-        self.assertEqual(self.presenter.view.set_value.call_count, 2)
+    def test_memory_usage_is_updated_based_on_a_constant(self):
```
Comment (martyngigg) [inline qt/python/mantidqt/widgets/memorywidget/test/test_memorypresenter.py:51]: I don't quite follow what the test name means here? Tests that depend on timings make me nervous. Can we atleast import the `TIME_INTERVAL_MEMORY_USAGE_UPDATE` constant from the module and add something to it?

Comment (Pasarus) [reply qt/python/mantidqt/widgets/memorywidget/test/test_memorypresenter.py:45]: Tests that the thread is doing what it is supposed to do, the variable is now used!

**Review by martyngigg — APPROVED** [review]
